x/pairing/keeper/unresponsive_provider.go

```diff
@@ -73,42 +74,70 @@ func (k Keeper) UnstakeUnresponsiveProviders(ctx sdk.Context, epochsNumToCheckCU
 	existingProviders := map[string]uint64{}
+	stakeAppliedBlockProviders := map[string]uint64{}
 	for _, providerStakeStorage := range providerStakeStorageList {
 		providerStakeEntriesForChain := providerStakeStorage.GetStakeEntries()
 		// count providers per geolocation
 		for _, providerStakeEntry := range providerStakeEntriesForChain {
 			if !providerStakeEntry.IsFrozen() {
 				existingProviders[providerStakeEntry.GetChain()]++
+				stakeAppliedBlockProviders[providerStakeEntry.Address] = providerStakeEntry.StakeAppliedBlock
 			}
 		}
 	}
-	// Go over the staked provider entries (on all chains)
-	for _, providerStakeStorage := range providerStakeStorageList {
-		providerStakeEntriesForChain := providerStakeStorage.GetStakeEntries()
-		for _, providerStakeEntry := range providerStakeEntriesForChain {
-			if minHistoryBlock < providerStakeEntry.StakeAppliedBlock {
-				// this staked provider has too short history (either since staking
-				// or since it was last unfrozen) - do not consider for jailing
-				continue
-			}
-			// update the CU count for this provider in providerCuCounterForUnreponsivenessMap
-			providerPaymentStorageKeyList, complaintCU, servicedCU, err := k.countCuForUnresponsiveness(ctx, minPaymentBlock, epochsNumToCheckCUForUnresponsiveProvider, epochsNumToCheckCUForComplainers, providerStakeEntry)
-			if err != nil {
-				utils.LavaFormatError("unstake unresponsive providers failed to count CU", err,
-					utils.Attribute{Key: "provider", Value: providerStakeEntry.Address},
+
+	// Go over the staked provider entries (on all chains) that has complaints
+	// build a map that has all the relevant details: provider address, chain, epoch and ProviderEpochCu object
+	pecsDetailed := k.GetAllProviderEpochCuStore(ctx)
+	complainedProviders := map[string]map[uint64]types.ProviderEpochCu{} // map[provider chainID]map[epoch]ProviderEpochCu
+	for _, pec := range pecsDetailed {
+		if minHistoryBlock < stakeAppliedBlockProviders[pec.Provider] {
+			// this staked provider has too short history (either since staking
+			// or since it was last unfrozen) - do not consider for jailing
+			continue
+		}
+		if _, ok := complainedProviders[pec.Provider+" "+pec.ChainId]; !ok {
+			complainedProviders[pec.Provider+" "+pec.ChainId] = map[uint64]types.ProviderEpochCu{pec.Epoch: pec.ProviderEpochCu}
+		} else {
+			if _, ok := complainedProviders[pec.Provider+" "+pec.ChainId][pec.Epoch]; !ok {
+				complainedProviders[pec.Provider+" "+pec.ChainId][pec.Epoch] = pec.ProviderEpochCu
+			} else {
+				utils.LavaFormatError("duplicate ProviderEpochCu key", fmt.Errorf("did not aggregate complainers CU"),
+					utils.LogAttr("key", types.ProviderEpochCuKey(pec.Epoch, pec.Provider, pec.ChainId)),
 				)
 				continue
 			}
+		}
+	}
+
+	// sort the keys so the iteration on the map will be deterministic
+	iterationOrder := []string{}
+	for key := range complainedProviders {
```

Comment:
forgot to sort?

Reply:
yes. good catch. see https://github.com/lavanet/lava/commit/60bdeb00a8b5a0052cea926f51676a44558299e9